### ISSUE TITLE
Snappy compression writer fails on certain encodings

### ISSUE THREAD

**Issue by dguidara** [elastacloud/parquet-dotnet#315]
**Version:** Parquet.Net v...
**Runtime Version:** .Net Framework v, .Net Core v etc.
**OS:** Windows/Linux v

#### Expected behavior

When I call      Schema schema = ParquetConvert.Serialize(structures, ms);
and the later on call 
            SimpleStructure[] structures2 = ParquetConvert.Deserialize<SimpleStructure>(ms);
I expect the objects to round trip correctly. 

#### Actual behavior

I instead get
threw exception: 
System.IO.EndOfStreamException: Unable to read beyond the end of the stream.

If I explicitly set compression mode to none, it works fine. 

### Steps to reproduce the behavior

1. Step 1.
2. Step 2.

#### Code snippet reproducing the behavior

      public void TestMethod1()
      {
         DateTime now = DateTime.Now;

         SimpleStructure[] structures = Enumerable
            .Range(0, 10)
            .Select(i => new SimpleStructure
            {
               Id = i,
               NullableId = (i % 2 == 0) ? new int?() : new int?(i),
               Name = $"row {i}",
               Date = now.AddDays(i).ToUniversalTime()
            })
            .ToArray();

         using (var ms = new MemoryStream())
         {
            Schema schema = ParquetConvert.Serialize(structures, ms);

            ms.Position = 0;

            SimpleStructure[] structures2 = ParquetConvert.Deserialize<SimpleStructure>(ms);

            for (int i = 0; i < 10; i++)
            {
             //  Assert.AreEqual(structures[i].Id, structures2[i].Id);
             //  Assert.AreEqual(structures[i].NullableId, structures2[i].NullableId);
               Assert.AreEqual(structures[i].Name, structures2[i].Name);
            //Assert.AreEqual(structures[i].Date, structures2[i].Date);
            }

         }
      }

            public class SimpleStructure
      {
         public int Id { get; set; }

         public int? NullableId { get; set; }

         public string Name { get; set; }

         public DateTimeOffset Date { get; set; }
      }
//here
```

- [ ] I'm attaching a sample .parquet file containing the problem.
- [ ] I am happy to contribute to this repository by fixing this issue and raising a Pull Request (see [contribution guide](CONTRIBUTING.md))

**Comment by dguidara** [comment]
Seems to be an issue with just Snappy compression.

     [Theory]
      [InlineData(CompressionMethod.None)]
      [InlineData(CompressionMethod.Gzip)]
      [InlineData(CompressionMethod.Snappy)]
      public void All_compression_methods_supported_string(CompressionMethod compressionMethod)
      {
         const string value = "test";
         object actual = WriteReadSingle(new DataField<string>("id"), value, compressionMethod);
         Assert.Equal(value, (string)actual);
      }

**Comment by aloneguid** [comment]
thanks @dguidara I can reproduce this now, investigating a fix

**Comment by tpeplow** [comment]
I see this too, this reproduces the issue (if you use None or Gzip it works)

```
var path = Path.Combine(Environment.CurrentDirectory, Guid.NewGuid() + ".parquet");
using (var file = File.Create(path))
{
var schema = new Schema(new DataField<string>("field"));
using (var parquetWriter = new ParquetWriter(schema, file) { CompressionMethod = CompressionMethod.Snappy })
using (var rowGroup = parquetWriter.CreateRowGroup(1))
{
    rowGroup.Write(new DataColumn((DataField) schema.Fields[0], new[] {"value 1"}));
}
}
using (var file = File.Open(path, FileMode.Open))
{
using (var parquetReader = new ParquetReader(file))
{
    using (var rowGroup = parquetReader.OpenRowGroupReader(0))
    {
            var col = rowGroup.ReadColumn(new DataField<string>("field"));
            ((string[]) col.Data).Single().ShouldEqual("value 1");
    }
}
}
```

**Comment by aloneguid** [comment]
We're experiencing this issue and it's a big pain, I'm all on it, please bear with me ;) Snappy library doesn't seems to be an issue, something Parquet library is doing is terribly wrong.

**Comment by aloneguid** [comment]
Found the issue, the fix is coming today.

**Comment by aloneguid** [comment]
This was due to the way paquet.net handles snappy streams. It would wrongly assume that when `.Flush()` is called for snappy, the stream is done, however certain encodings will call Flush internally, therefore result file would end up in partially completed snappy stream. The fix is available on NuGet as v3.0.1.
